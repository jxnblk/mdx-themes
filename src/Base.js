// vanilla css theme
import React from 'react'
import { MDXProvider } from '@mdx-js/tag'
import baseComponents from './components'
import CSSVars from './CSSVars'

const baseTheme = {
  h1: '3rem',
  h2: '2.25rem',
  h3: '1.5rem',
  h4: '1.125rem',
  h5: '1rem',
  h6: '.75rem',
  lineHeight: 4/3,
  maxWidth: '1024px',
  m1: 2/3 + 'em',
  m2: 4/3 + 'em',
  m3: 8/3 + 'em',
  m4: 16/3 + 'em',
  color: 'black',
  background: 'white',
  link: '#07c',
  pre: '#000',
  preBackground: '#f6f8fa',
  radius: '2px',
}

const css = `
.mdx-base-theme {
  line-height: var(--line-height);
  max-width: var(--max-width);
  margin: auto;
  padding: var(--m2);
  color: var(--color);
  background-color: var(--background);
}

.mdx-a {
  color: var(--link);
}

.mdx-img {
  max-width: 100%;
  height: auto;
}

.mdx-h1 { font-size: var(--h1) }
.mdx-h2 { font-size: var(--h2) }
.mdx-h3 { font-size: var(--h3) }
.mdx-h4 { font-size: var(--h4) }
.mdx-h5 { font-size: var(--h5) }
.mdx-h6 { font-size: var(--h6) }

.mdx-h1, .mdx-h2, .mdx-h3 {
  margin-top: var(--m1);
  margin-bottom: 0;
}

.mdx-h4, .mdx-h5, .mdx-h6,
.mdx-p, .mdx-ol, .mdx-ul,
.mdx-img, .mdx-blockquote {
  margin-top: var(--m2);
  margin-bottom: var(--m2);
}

.mdx-pre, .mdx-code {
  font-family: Menlo, monospace;
  font-size: var(--h5);
}

.mdx-pre {
  padding: var(--m1);
  color: var(--pre);
  background-color: var(--pre-background);
  border-radius: var(--radius);
}

.mdx-blockquote {
  font-size: var(--h3);
  font-weight: bold;
}
`

const style = (
  <style
    dangerouslySetInnerHTML={{
      __html: css
    }}
  />
)

export const BaseTheme = ({
  theme = {},
  components = {},
  children
}) =>
  <CSSVars
    className='mdx-base-theme'
    theme={{
      ...baseTheme,
      ...theme
    }}>
    <MDXProvider
      components={{
        ...baseComponents,
        ...components
      }}>
      <React.Fragment>
        {style}
        {children}
      </React.Fragment>
    </MDXProvider>
  </CSSVars>

