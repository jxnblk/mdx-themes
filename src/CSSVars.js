import React from 'react'

const hyphenate = s => s.replace(/[A-Z]|^ms/g, '-$&').toLowerCase()

export const createStyle = theme => {
  const style = {}
  for (const key in theme) {
    style['--' + hyphenate(key)] = theme[key]
  }
  return style
}

export default ({
  theme = {},
  className,
  children
}) => {
  const style = createStyle(theme)

  return (
    <div
      className={className}
      style={style}>
      {children}
    </div>
  )
}
