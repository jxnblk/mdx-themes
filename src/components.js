// vanilla css components
import React from 'react'
import { Link as RouterLink } from 'react-router-dom'
import isAbsolute from 'is-absolute-url'

export const Link = ({
  href,
  ...props
}) => isAbsolute(href)
  ? <a {...props} href={href} />
  : <RouterLink {...props} to={href} />

const join = (...args) => args.filter(Boolean).join(' ')
const classname = (Tag, className) => props =>
  <Tag {...props} className={join(props.className, className)} />

const heading = Tag => ({ id, children, ...props }) =>
  <Tag {...props} id={id}>
    <a
      href={'#' + id}
      style={{
        textDecoration: 'none',
        color: 'inherit'
      }}>
      {children}
    </a>
  </Tag>

const h1 = classname(heading('h1'), 'mdx-h1')
const h2 = classname(heading('h2'), 'mdx-h2')
const h3 = classname(heading('h3'), 'mdx-h3')
const h4 = classname(heading('h4'), 'mdx-h4')
const h5 = classname(heading('h5'), 'mdx-h5')
const h6 = classname(heading('h6'), 'mdx-h6')
const a = classname(Link, 'mdx-a')
const p = classname('p', 'mdx-p')
const ul = classname('ul', 'mdx-ul')
const ol = classname('ol', 'mdx-ol')
const li = classname('li', 'mdx-li')
const blockquote = classname('blockquote', 'mdx-blockquote')
const img = classname('img', 'mdx-img')
const pre = classname('pre', 'mdx-pre')
const code = classname('code', 'mdx-code')

export default {
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  a,
  p,
  ol,
  ul,
  li,
  blockquote,
  img,
  // adjust for remark
  pre: props => props.children,
  code: pre,
  inlineCode: code,
}
