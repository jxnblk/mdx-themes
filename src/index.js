export { BaseTheme } from './Base'
export { default as components } from './components'
export { default as CSSVars } from './CSSVars'
