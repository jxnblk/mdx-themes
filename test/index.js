import React from 'react'
import TestRenderer from 'react-test-renderer'
import { BaseTheme } from '../src'

const renderJSON = el => TestRenderer.create(el).toJSON()

describe('BaseTheme', () => {
  test('renders', () => {
    const json = renderJSON(
      <BaseTheme />
    )
    expect(json).toMatchSnapshot()
  })
})
