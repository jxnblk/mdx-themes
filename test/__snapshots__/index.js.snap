// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTheme renders 1`] = `
<div
  className="mdx-base-theme"
  style={
    Object {
      "--background": "white",
      "--color": "black",
      "--h1": "3rem",
      "--h2": "2.25rem",
      "--h3": "1.5rem",
      "--h4": "1.125rem",
      "--h5": "1rem",
      "--h6": ".75rem",
      "--line-height": 1.3333333333333333,
      "--link": "#07c",
      "--m1": "0.6666666666666666em",
      "--m2": "1.3333333333333333em",
      "--m3": "2.6666666666666665em",
      "--m4": "5.333333333333333em",
      "--max-width": "1024px",
      "--pre": "#000",
      "--pre-background": "#f6f8fa",
      "--radius": "2px",
    }
  }
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "
.mdx-base-theme {
  line-height: var(--line-height);
  max-width: var(--max-width);
  margin: auto;
  padding: var(--m2);
  color: var(--color);
  background-color: var(--background);
}

.mdx-a {
  color: var(--link);
}

.mdx-img {
  max-width: 100%;
  height: auto;
}

.mdx-h1 { font-size: var(--h1) }
.mdx-h2 { font-size: var(--h2) }
.mdx-h3 { font-size: var(--h3) }
.mdx-h4 { font-size: var(--h4) }
.mdx-h5 { font-size: var(--h5) }
.mdx-h6 { font-size: var(--h6) }

.mdx-h1, .mdx-h2, .mdx-h3 {
  margin-top: var(--m1);
  margin-bottom: 0;
}

.mdx-h4, .mdx-h5, .mdx-h6,
.mdx-p, .mdx-ol, .mdx-ul,
.mdx-img, .mdx-blockquote {
  margin-top: var(--m2);
  margin-bottom: var(--m2);
}

.mdx-pre, .mdx-code {
  font-family: Menlo, monospace;
  font-size: var(--h5);
}

.mdx-pre {
  padding: var(--m1);
  color: var(--pre);
  background-color: var(--pre-background);
  border-radius: var(--radius);
}

.mdx-blockquote {
  font-size: var(--h3);
  font-weight: bold;
}
",
      }
    }
  />
</div>
`;
